pypose/module/pf.py

```diff
@@ -0,0 +1,204 @@
+        ...     v = r * torch.randn(N) # observation noise
+        ...     states[i+1], observ[i] = model(states[i] + w, inputs[i])
+        ...     estim[i+1], P[i+1] = pf(estim[i], observ[i] + v, inputs[i], P[i], Q, R)
+        ... print('Est error:', (states - estim).norm(dim=-1))
+        Est error: tensor([5.3627, 0.5640, 0.0953, 0.0447, 0.0936])
+
+    Note:
+        Implementation is based on Section 15.2 of this book
+
+        * Dan Simon, `Optimal State Estimation: Kalman, H∞, and Nonlinear Approaches
+          <https://onlinelibrary.wiley.com/doi/epdf/10.1002/0470045345.fmatter>`_,
+          Cleveland State University, 2006
+    '''
+
+    def __init__(self, model, Q=None, R=None, particle_number=None):
+        super().__init__(model, Q, R)
+        self.particle_number = 1000 if particle_number is None else particle_number
+
+    def forward(self, x, y, u, P, Q=None, R=None, t=None, conv_weight=None):
+        r'''
+        Performs one step estimation.
+
+        Args:
+            x (:obj:`Tensor`): estimated system state of previous step
+            y (:obj:`Tensor`): system observation at current step (measurement)
+            u (:obj:`Tensor`): system input at current step
+            P (:obj:`Tensor`): state estimation covariance of previous step
+            Q (:obj:`Tensor`, optional): covariance of system transition model
+            R (:obj:`Tensor`, optional): covariance of system observation model
+
+        Return:
+            list of :obj:`Tensor`: posteriori state and covariance estimation
+        '''
+        # Upper cases are matrices, lower cases are vectors
+
+        Q = Q if Q is not None else self.Q
+        R = R if R is not None else self.R
+        conv_weight = len(x) if conv_weight is None else conv_weight
+        self.model.set_refpoint(state=x, input=u, t=t)
+
+        xp = self.generate_particle(x, conv_weight * P)
+        xs = self.model.state_transition(xp, u, t)
+        ye = self.model.observation(xs, u, t)
+        q = self.relative_likelihood(y, ye, R)
+        xr = self.resample_particles(q, xs)
+
+        x = xr.mean(dim=0)
+        ex = xr - x
+        weight = torch.tensor([1 / self.particle_number])
+        P = self.compute_cov(ex, ex, weight, Q)
+
+        return x, P
+
+    def generate_particle(self, x, P):
+        r'''
+        Randomly generate particles
+        '''
+        xp = torch.distributions.MultivariateNormal(x, P).sample(
+            (self.particle_number,))
+        return xp
+
+    def relative_likelihood(self, y, ye, R):
+        r'''
+        Compute the relative likelihood
+        '''
+        q = torch.distributions.MultivariateNormal(ye, R).log_prob(y).exp()
+        q = q / torch.sum(q)
+        return q
+
+    def resample_particles(self, q, x):
+        r'''
+        Resample the set of a posteriori particles
+        '''
+        r = torch.rand(self.particle_number, device=x.device)
+        cumsumq = torch.cumsum(q, dim=0)
+        cumsumq[-1] = 1.0
```

Comment:
do you want `torch.testing.assert_allclose(cumsumq[-1], torch.tensor(1.0, device=r.device))`

Reply:
> assert_allclose
**log_prob.exp()** will get the **None** list, we  need to assign -1 to avoid the denominator to be None
```python
 q = MultivariateNormal(ye, R).log_prob(y).exp()
```